File: weekly-contest/184/4.go
package main

import "fmt"

func numOfWays(n int) int {
	mod := 1000000007
	same, diff := 6, 6
	for i := 1; i < n; i++ {
		nextSame := 3*same + 2*diff
		nextDiff := 2*same + 2*diff
		same, diff = (nextSame % mod), (nextDiff % mod)
	}

	return (same + diff) % mod
}

func main() {
	fmt.Println(numOfWays(1))
	fmt.Println(numOfWays(2))
	fmt.Println(numOfWays(3))
	fmt.Println(numOfWays(7))
	fmt.Println(numOfWays(5000))
}

File: weekly-contest/184/4.rb
# @param {Integer} n
# @return {Integer}
def num_of_ways(n)
  same, diff = 6, 6
  mod = 10**9 + 7
  (n-1).times do |i|
    n_s = 3 * same + 2 * diff
    n_d = 2 * same + 2 * diff

    same, diff = n_s%mod, n_d%mod
  end

  (same + diff)%mod
end

# p num_of_ways(1)
# p num_of_ways(2)
# p num_of_ways(3)
p num_of_ways(5000)